pkg/clientcert/cert_controller.go

```diff
@@ -0,0 +1,350 @@
+package clientcert
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509/pkix"
+	"fmt"
+	"math/rand"
+	"reflect"
+	"time"
+
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+
+	certificates "k8s.io/api/certificates/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	certificatesinformers "k8s.io/client-go/informers/certificates/v1beta1"
+	corev1informers "k8s.io/client-go/informers/core/v1"
+	csrclient "k8s.io/client-go/kubernetes/typed/certificates/v1beta1"
+	corev1client "k8s.io/client-go/kubernetes/typed/core/v1"
+	certificateslisters "k8s.io/client-go/listers/certificates/v1beta1"
+	cache "k8s.io/client-go/tools/cache"
+	certutil "k8s.io/client-go/util/cert"
+	"k8s.io/client-go/util/keyutil"
+	"k8s.io/klog/v2"
+)
+
+const (
+	// KubeconfigFile is the name of the kubeconfig file in kubeconfigSecret
+	KubeconfigFile = "kubeconfig"
+	// TLSKeyFile is the name of tls key file in kubeconfigSecret
+	TLSKeyFile = "tls.key"
+	// TLSCertFile is the name of the tls cert file in kubeconfigSecret
+	TLSCertFile = "tls.crt"
+
+	clusterNameAnnotation = "open-cluster-management.io/cluster-name"
+	ClusterNameFile       = "cluster-name"
+	AgentNameFile         = "agent-name"
+
+	ClusterNameLabel = "open-cluster-management.io/cluster-name"
+	AddonNameLabel   = "open-cluster-management.io/addon-name"
+	SignerNameLabel  = "open-cluster-management.io/signer-name"
+)
+
+// ControllerResyncInterval is exposed so that integration tests can crank up the constroller sync speed.
+var ControllerResyncInterval = 5 * time.Minute
+
+// CSROption includes options that is used to create csrs
+type CSROption struct {
+	// ObjectMeta is the ObjectMeta shared by all created csrs. It should use GenerateName instead of Name
+	// to generate random csr names
+	ObjectMeta metav1.ObjectMeta
+	// Subject represents the subject of the client certificate used to create csrs
+	Subject *pkix.Name
+	// SignerName is the name of the signer specified in the created csrs
+	SignerName string
+}
+
+// ClientCertOption includes options that is used to create client certificate
+type ClientCertOption struct {
+	// SecretNamespace is the namespace of the secret containing client certificate.
+	SecretNamespace string
+	// SecretName is the name of the secret containing client certificate. The secret will be created if
+	// it does not exist.
+	SecretName string
+	// AdditonalSecretData contains data that will be added into client certificate secret besides tls.key/tls.crt
+	AdditonalSecretData map[string][]byte
```

Comment:
i think csrFilterFunc should also be put here?

Reply:
Updated as suggested.